cmd/juju/charmhub/printinfo.go

```diff
@@ -0,0 +1,535 @@
+	return wrapFlow(iw.w, quotedIfNeeded(iw.in.Charm.Summary), "summary:     ", iw.termWidth)
+}
+
+func (iw *infoWriter) maybePrintPublisher() error {
+	publisher, ok := iw.in.Charm.Publisher["display-name"]
+	if !ok {
+		return nil
+	}
+	_, err := fmt.Fprintf(iw.w, "publisher:   %s\n", publisher)
+	return err
+}
+
+func (iw *infoWriter) maybePrintDescription() error {
+	if iw.in.Charm.Description == "nil" {
+		return nil
+	}
+
+	if len(iw.in.Charm.Description) < iw.termWidth-13 {
+		_, err := fmt.Fprintf(iw.w, "description: %s\n", iw.in.Charm.Description)
+		return err
+	}
+
+	if _, err := fmt.Fprintln(iw.w, "description: |"); err != nil {
+		return err
+	}
+
+	return printDescr(iw.w, iw.in.Charm.Description, iw.termWidth)
+}
+
+func (iw *infoWriter) maybePrintStoreURL() error {
+	// TODO (hml) 2020-06-24
+	// implement once the data is available.
+	return nil
+}
+
+func (iw *infoWriter) maybePrintChannels() error {
+	// TODO (hml) 2020-06-24
+	// Implement up arrow for closed channels
+	tw := output.TabWriter(iw.w)
+	w := output.Wrapper{tw}
+	if len(iw.in.ChannelMap) == 0 {
+		return nil
+	}
+	if _, err := fmt.Fprintln(iw.w, "channels:"); err != nil {
+		return err
+	}
+	for _, ch := range iw.in.ChannelMap {
+		releasedAt, err := time.Parse(time.RFC3339, ch.Channel.ReleasedAt)
+		if err != nil {
+			return errors.Annotate(err, "could not parse released at time")
+		}
+
+		w.Printf("  %s:", ch.Channel.Name)
+		w.Print(ch.Revision.Version)
+		w.Print(releasedAt.Format("2006-01-02"))
+		w.Printf("(%s)", strconv.Itoa(ch.Revision.Revision))
+		w.Println(sizeToStr(ch.Revision.Download.Size))
+	}
+	return tw.Flush()
+}
+
+type bundleInfoWriter struct {
+	*infoWriter
+}
+
+func (b *bundleInfoWriter) maybePrintID() error {
+	if b.in.ID == "" {
+		return nil
+	}
+	_, err := fmt.Fprintf(b.w, "bundle-id:   %s\n", b.in.ID)
+	return err
+}
+
+func (b *bundleInfoWriter) Print() error {
+	// Order is important here, to keep the output
+	// the same.
+	if err := b.maybePrintName(); err != nil {
+		return nil
+	}
+	if err := b.maybePrintID(); err != nil {
+		return nil
+	}
+	if err := b.maybePrintSummary(); err != nil {
+		return nil
+	}
+	if err := b.maybePrintPublisher(); err != nil {
+		return nil
+	}
+	if err := b.maybePrintStoreURL(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := b.maybePrintDescription(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := b.maybePrintCharms(); err != nil {
+		return errors.Trace(err)
+	}
+	if err := b.maybePrintChannels(); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (b *bundleInfoWriter) maybePrintCharms() error {
+	// TODO (hml) 2020-06-25
+	// Implement once the data is available to find.
+	return nil
+}
+
+type charmInfoWriter struct {
+	*infoWriter
+	charmMeta   *charm.Meta
+	charmConfig *charm.Config
+}
+
+func (c *charmInfoWriter) Print() error {
+	c.maybeUnmarshalCharmMetadata()
+	c.maybeUnmarshalCharmConfig()
+	// Order is important here, to keep the output
+	// the same.
```

Comment:
This should totally be a visitor pattern and that way the order is made explicit by the order of the slice.

```go
printers := []Printer{Name{}, ID{}, Summary{}, Publisher{}}
for _, printer := range printers {
    if err := printer.Print(w); err != nil {
        return errors.Trace(err)
    }
}
```